### DISCUSSION TITLE
The integrated showing error

### DISCUSSION THREAD

**Discussion by honghh2018** [brianhie/scanorama#96] [Q&A]
Hi guy,
 When i runing on below step,
adatas_cor = scanorama.correct_scanpy(adatas, return_dimred=True) 
 It was complain about error:
Error: No genes found in all datasets, exiting...
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
/tmp/ipykernel_146881/3261772617.py in <module>
----> 1 adatas_cor = scanorama.correct_scanpy(adatas, return_dimred=True)

/share/nas1/Data/software/mini_conda/Miniconda3/envs/Scanpy_package/lib/python3.8/site-packages/scanorama/scanorama.py in correct_scanpy(adatas, **kwargs)
    219     """
    220     if 'return_dimred' in kwargs and kwargs['return_dimred']:
--> 221         datasets_dimred, datasets, genes = correct(
    222             [adata.X for adata in adatas],
    223             [adata.var_names.values for adata in adatas],

/share/nas1/Data/software/mini_conda/Miniconda3/envs/Scanpy_package/lib/python3.8/site-packages/scanorama/scanorama.py in correct(datasets_full, genes_list, return_dimred, batch_size, verbose, ds_names, dimred, approx, sigma, alpha, knn, return_dense, hvg, union, seed)
     90     datasets_full = check_datasets(datasets_full)
     91 
---> 92     datasets, genes = merge_datasets(datasets_full, genes_list,
     93                                      ds_names=ds_names, union=union)
     94     datasets_dimred, genes = process_data(datasets, genes, hvg=hvg,

/share/nas1/Data/software/mini_conda/Miniconda3/envs/Scanpy_package/lib/python3.8/site-packages/scanorama/scanorama.py in merge_datasets(datasets, genes, ds_names, verbose, union)
    336         if len(keep_genes) == 0:
    337             print('Error: No genes found in all datasets, exiting...')
--> 338             exit(1)
    339     if verbose:
    340         print('Found {} genes among all datasets'

NameError: name 'exit' is not defined

So how can i fix this issue?
Any advice would be appreciated

**Comment by brianhie** [comment]
Hi @honghh2018, could you make sure that the genes you are passing to scanorama are the same? By default, it will take the intersection of all gene sets.

**Reply by honghh2018** [reply]
So, how can i add intersection of the  two samples genes in  adatas just like i post above object.
Best,
hanhuihong

**Reply by brianhie** [reply]
You can pass in the gene sets as adatas. Just make sure that
```python
len(set(adata1.var_names) & set(adata2.var_names)) > 0
```

**Reply by honghh2018** [reply]
You mean like below, using the intersection genes that within in two adata objects? the command lying below:
    adatas_cor = scanorama.correct_scanpy(adatas, return_dimred=True, genes[set(adata1.var_names) & set(adata2.var_names)])

I would try and feedback soon.
Best

**Reply by honghh2018** [reply]
sorry for the late reply,
I found the something weird that vignette post in you homepage about data integrated by scanorama.
https://nbisweden.github.io/workshop-scRNAseq/labs/compiled/scanpy/scanpy_03_integration.html
It was too complicated to figure it out myself.  

in my code, i want to integrated two single cell sample by scanorama algorithm was processing into below step:
STIM=sc.read_csv("./t2_STIM.xls").T 
ctrl=sc.read_csv("./t2_ctrl.xls").T 
STIM.var_names_make_unique()
ctrl.var_names_make_unique()
sc.pp.calculate_qc_metrics(STIM, inplace=True)
sc.pp.calculate_qc_metrics(ctrl, inplace=True)
for adata in [
    STIM,
    ctrl,
]:
    sc.pp.normalize_total(adata, inplace=True)
    sc.pp.log1p(adata)
    sc.pp.highly_variable_genes(adata, flavor="seurat", n_top_genes=2000, inplace=True)


adatas = [STIM, ctrl]
adatas_cor = scanorama.correct_scanpy(adatas, return_dimred=True)

and then, I can real not how to step next when i process my datas come at this step:  
adatas_cor = scanorama.correct_scanpy(adatas, return_dimred=True)
Would you please give me some tips to kick this issue off?
Best,
Hanhuihong

**Reply by honghh2018** [reply]
@brianhie 
Had any advice about my issue?
Is that right about my code?
Best
